src/dotnet/MLSearch.Service/Bot/Tools/Context/BotToolsContextHandler.cs

```diff
@@ -0,0 +1,95 @@
+using System.IdentityModel.Tokens.Jwt;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.Tokens;
+
+namespace ActualChat.MLSearch.Bot.Tools.Context;
+public class BotToolsContextHandlerOptions {
+    public SigningCredentials SigningCredentials{ get; set; }
+    public string Audience { get; set; }
+    public string Issuer { get; set; }
+    public TimeSpan ContextLifetime { get; set; }
+}
+
+public class BotToolsContext(ClaimsPrincipal? claims): IBotToolsContext {
+    public const string ConversationClaimType = "ConversationId";
+    public string? ConversationId => claims?.FindFirstValue(ConversationClaimType);
+    public bool IsValid => claims != null 
+        && claims.HasClaim(e => string.Equals(e.Type, ConversationClaimType, StringComparison.Ordinal));
+    
+}
+
+public class BotToolsContextHandler(IOptionsMonitor<BotToolsContextHandlerOptions> options): IBotToolsContextHandler
+{
+    private const string BEARER_PREFIX = "Bearer ";
+
+    public IBotToolsContext GetContext(HttpRequest request) => new BotToolsContext(GetValidatedClaims(request));
+    public void SetContext(HttpRequestMessage request, string conversationId) {
+        var config = options.CurrentValue;
+        var claims = new List<Claim>();
+        claims.Add(new Claim(BotToolsContext.ConversationClaimType, conversationId));
+        var expiration = DateTime.Now.Add(config.ContextLifetime);
+
+        var token = new JwtSecurityToken(
+            issuer: config.Issuer,
+            audience: config.Audience,
+            claims: claims,
+            expires: expiration,
+            signingCredentials: config.SigningCredentials
+        );
+
+        var signedToken = new JwtSecurityTokenHandler().WriteToken(token);
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", signedToken);
+    }
+
+    private ClaimsPrincipal? GetValidatedClaims(HttpRequest request) {
+        string? bearerToken = request.Headers.Authorization
+            .Where(e => e?.StartsWith(BEARER_PREFIX, StringComparison.InvariantCulture) == true)
```

Comment:
StringComparison.Ordinal is faster than InvariantCulture

Reply:
ok